signer/pkcs11/pkcs11key/pkcs11key.go

```diff
@@ -0,0 +1,235 @@
+	if err != nil {
+		ps.Destroy()
+		return
+	}
+	if err = p.FindObjectsFinal(session); err != nil {
+		ps.Destroy()
+		return
+	}
+
+	if len(objs) == 0 {
+		err = errors.New("private key not found")
+		ps.Destroy()
+		return
+	}
+	ps.privateKeyHandle = objs[0]
+
+	// Populate the pubic key from the private key
+	// TODO: Add support for non-RSA keys, switching on CKA_KEY_TYPE
+	template = []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_MODULUS, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_PUBLIC_EXPONENT, nil),
+	}
+	attr, err := p.GetAttributeValue(session, ps.privateKeyHandle, template)
+	if err != nil {
+		ps.Destroy()
+		return
+	}
+
+	n := big.NewInt(0)
+	e := int(0)
+	gotModulus, gotExponent := false, false
+	for _, a := range attr {
+		if a.Type == pkcs11.CKA_MODULUS {
+			n.SetBytes(a.Value)
+			gotModulus = true
+		} else if a.Type == pkcs11.CKA_PUBLIC_EXPONENT {
+			bigE := big.NewInt(0)
+			bigE.SetBytes(a.Value)
+			e = int(bigE.Int64())
+			gotExponent = true
+		}
+	}
+	if !gotModulus || !gotExponent {
+		ps.Destroy()
+		return
+	}
+	ps.publicKey = rsa.PublicKey{n, e}
+
+	return
+}
+
+// Destroy tears down a PKCS11Key.
+//
+// This method must be called before the PKCS11Signer is GC'ed, in
+// order to ensure that the PKCS#11 module itself is properly finalized
+// and destroyed.
+//
+// The idiomatic way to do this (assuming no need for a long-lived
+// signer) is as follows:
+//
+//   ps, err := NewPKCS11Signer(...)
+//   if err != nil { ... }
+//   defer ps.Destroy()
+func (ps *PKCS11Key) Destroy() {
+	if ps.module != nil {
+		ps.module.Finalize()
+		ps.module.Destroy()
+	}
+}
+
+func (ps *PKCS11Key) openSession() (session pkcs11.SessionHandle, err error) {
+	// Find slot by description
+	slots, err := ps.module.GetSlotList(true)
+	if err != nil {
+		return
+	}
+	for _, slot := range slots {
+		slotInfo, err := ps.module.GetSlotInfo(slot)
+		if err != nil {
+			continue
+		}
+
+		if slotInfo.SlotDescription == ps.slotDescription {
+			// Open session
+			session, err = ps.module.OpenSession(slot, pkcs11.CKF_SERIAL_SESSION)
+			if err != nil {
+				return session, err
+			}
+
+			// Login
+			if err = ps.module.Login(session, pkcs11.CKU_USER, ps.pin); err != nil {
+				return session, err
+			}
+
+			return session, err
+		}
+	}
+
+	err = errors.New("slot not found")
+	return
+}
+
+func (ps *PKCS11Key) closeSession(session pkcs11.SessionHandle) {
+	ps.module.Logout(session)
+	ps.module.CloseSession(session)
+}
+
+// Public returns the public key for the PKCS #11 key.
+func (ps *PKCS11Key) Public() crypto.PublicKey {
+	return &ps.publicKey
+}
+
+// Sign performs a signature using the PKCS #11 key.
+func (ps *PKCS11Key) Sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	// Verify that the length of the hash is as expected
+	hash := opts.HashFunc()
+	hashLen := hash.Size()
+	if len(msg) != hashLen {
+		err = errors.New("input size does not match hash function output size")
```

Comment:
Should these errors switch to using cferr?  Actually probably not, since it might be nice to be able to re-use this file outside of CFSSL.